File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        triplets = []
        for i in range(len(nums) - 2):
            if(i > 0 and nums[i] == nums[i-1]):
                continue
            j = i+1
            k = len(nums) - 1

            while(j < k):
                threesum = nums[i] + nums[j] + nums[k]
                if( threesum == 0):
                    triplets.append([nums[i], nums[j], nums[k]])
                    while(j < k and nums[j] == nums[j+1]):
                        j += 1
                    while(k > j and nums[k] == nums[k-1]):
                        k -= 1
                    j += 1
                    k -= 1
                elif(threesum > 0):
                    k -= 1
                else:
                    j += 1
        return triplets        

File: 15-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let triplets = []
    nums.sort((a,b) => a-b)

    for(let i=0; i<nums.length-2; i++){
        if(i > 0 && nums[i] == nums[i-1]) continue

        let p1 = i+1, p2 = nums.length-1

        while(p1 < p2){
            let sum = nums[i] + nums[p1] + nums[p2]

            if(sum == 0){
                triplets.push([nums[i], nums[p1], nums[p2]])
                p1++
                p2--

                while(p1 < p2 && nums[p1] == nums[p1-1]) p1++
                while(p1 < p2 && nums[p2] == nums[p2+1]) p2-- 
            }
            else if(sum < 0){
                p1++
            }else{
                p2--
            }
        }
    }
    return triplets
};